File: 0778-reorganize-string/0778-reorganize-string.py
class Solution:
    def reorganizeString(self, s: str) -> str:
        hash_freq=dict()
        for c in s:
            hash_freq[c]=hash_freq.get(c,0)+1
        #print(hash_freq)
        max_heap=[(-val,key) for key,val in hash_freq.items()]
        #print(max_heap)
        heapq.heapify(max_heap)
        #print(max_heap)
        result=""
        while len(max_heap)>1:
            f1,char1=heapq.heappop(max_heap)
            f2,char2=heapq.heappop(max_heap)
            result+=char1+char2
            if f1+1<0: 
                heapq.heappush(max_heap,(f1+1,char1)) 
            if f2+1<0:
                heapq.heappush(max_heap,(f2+1,char2)) 
            print(result)
        if max_heap:
            if -max_heap[0][0]>1:
                return ""
            result+=max_heap[0][1] 
        return result

File: 0778-reorganize-string/0778-reorganize-string.java
class Solution {
    public String reorganizeString(String s) {
        HashMap<Character,Integer> hash_freq=new HashMap<>();
        for(char c:s.toCharArray()){
            hash_freq.put(c,hash_freq.getOrDefault(c,0)+1);
        }
        String result="";
        PriorityQueue<Character> max_heap=new PriorityQueue<>((a,b)->hash_freq.get(b)-hash_freq.get(a));
        max_heap.addAll(hash_freq.keySet());
        while(max_heap.size()>1){
            char c1=max_heap.poll();
            char c2=max_heap.poll();
            result+=c1+""+c2;
            hash_freq.put(c1,hash_freq.get(c1)-1);
            hash_freq.put(c2,hash_freq.get(c2)-1);
            if(hash_freq.get(c1)>0) max_heap.add(c1);
            if(hash_freq.get(c2)>0) max_heap.add(c2);
        }
        if(!max_heap.isEmpty()){
            char c=max_heap.poll();
            if(hash_freq.get(c)>1) return "";
            result+=c;
        }
        return result;
    }
    
}